New Optimizer for multi receiver

nishanthmarer opened the pull request:
This contains the new optimizer code for multi receivers

nishanthmarer committed:
New Optimizer for multi receiver

SpontaneousDuck committed:
initial optimizer integration work

SpontaneousDuck commented:
I wrote initial integration code and am getting wildly incorrect results on first run. Using test code below, results are off by about 100dB and the optimizer fails: 
```python
from scenario_generator import ScenarioGenerator
import torch
batch_size = 128
n_bs = 1
n_ut = 1
h_ut = 1.5
h_bs = 10.0
resolution=500
seed=44
snr = 10.0

scenario_gen = ScenarioGenerator(n_receivers=n_bs, n_transmitters=n_ut, batch_size=batch_size, h_receivers=h_bs, h_transmitters=h_ut, resolution=resolution, min_receiver_dist=100, seed=seed, n_workers=1, dtype=torch.float32, device=torch.device("cuda:0"))
scenario_gen.RegenerateFullScenario(target_snr=snr)
```

SpontaneousDuck committed:
Appears to work for most cases now

SpontaneousDuck commented:
Code is updated to work with most cases now. We need to handle regenerating points where the solution is very unlikely. (estimated SNR is not close to required due to map edge constraints)

SpontaneousDuck committed:
Add total SNR calculation and debug flag

nishanthmarer committed:
1. Assign SNR to Receivers
2. Exception raised when max iteration is reached for finding optimal receivers
3. Included ortools as a requirement

nishanthmarer commented:
Pushed the updated code, containing the smart SNR assignment to receivers.
Exception raised for exceeding max iteration for finding the optimal receiver location
Included "ortools" in requirement list

SpontaneousDuck committed:
Add dependency on SionnaTorch

nishanthmarer committed:
1. Given a single SNR, it should be perceived as total SNR and distributed across the Rx Towers.
2. Given a list of SNR, treat as assignment to each Rx and obtain the optimal value for each Rx.

3. Added a flag which still allows for replication of SNRs across the RxTowers.
Flag Name: replicateSNR

nishanthmarer committed:
Edge case consideration for unachievable SNR

nishanthmarer committed:
Annotate the SNR along the Tx and Rx path.

Added a new requirement: matplotlib-label-lines

Needed for annotation

nishanthmarer committed:
Change from SNR to Linear Power.

SNR cannot be added, only Power can be added in the linear space

nishanthmarer committed:
Minor changes

nishanthmarer committed:
Change from Percentage Diff to RMS Error

SpontaneousDuck committed:
return SNR from scenario generator

nishanthmarer committed:
Temporary Memory Management fix
